lib/edgehog_device_forwarder/connections.ex

```diff
@@ -0,0 +1,147 @@
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#c:terminate/2
+    #   - https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html#module-custom-transports
+    ConCache.insert_new(@tokens, token, TokenData.new(socket))
+  end
+
+  @spec close_socket(token) :: :ok
+  def close_socket(token) do
+    ConCache.delete(@tokens, token)
+  end
+
+  @spec fetch_socket(token) :: {:ok, socket} | {:error, :token_not_found}
+  def fetch_socket(token) do
+    case ConCache.get(@tokens, token) do
+      nil -> {:error, :token_not_found}
+      %TokenData{socket: socket} -> {:ok, socket}
+    end
+  end
+
+  ## Messages
+
+  @spec new_http_request(controller) :: {:ok, http_request_id}
+  def new_http_request(controller) do
+    request_id = UUID.uuid4(:raw)
+
+    case ConCache.insert_new(@messages, request_id, {:http, controller}) do
+      :ok -> {:ok, request_id}
+      {:error, :already_exists} -> new_http_request(controller)
+    end
+  end
+
+  @spec fetch_http_request(http_request_id, :respond | {:upgrade, :websocket}) ::
+          {:ok, controller} | {:error, :http_request_not_found}
+  def fetch_http_request(request_id, action) do
+    ConCache.isolated(@messages, request_id, fn ->
+      case ConCache.get(@messages, request_id) do
+        {:http, controller} ->
+          # we already have the lock from the isolated/3 call
+          case action do
+            :respond -> ConCache.dirty_delete(@messages, request_id)
+            {:upgrade, :websocket} -> ConCache.dirty_put(@messages, request_id, :upgrading_to_ws)
+          end
+
+          {:ok, controller}
+
+        _ ->
+          {:error, :http_request_not_found}
+      end
+    end)
+  end
+
+  @spec upgrade_to_websocket(http_request_id, socket) ::
+          :ok | {:error, :http_request_not_found}
+  def upgrade_to_websocket(request_id, socket) do
+    ConCache.isolated(@messages, request_id, fn ->
+      case ConCache.get(@messages, request_id) do
+        :upgrading_to_ws ->
+          ConCache.dirty_put(@messages, request_id, {:ws, socket})
+
+        _ ->
+          {:error, :http_request_not_found}
+      end
+    end)
+  end
+
+  @spec fetch_target_websocket(ws_socket_id) ::
+          {:ok, socket} | {:error, :websocket_not_found}
+  def fetch_target_websocket(socket_id) do
+    case ConCache.get(@messages, socket_id) do
+      {:ws, socket} -> {:ok, socket}
+      _ -> {:error, :websocket_not_found}
+    end
+  end
+
+  @spec close_user_websocket(ws_socket_id) :: :ok | {:error, :websocket_not_found}
+  def close_user_websocket(socket_id) do
+    ConCache.isolated(@messages, socket_id, fn ->
+      case ConCache.get(@messages, socket_id) do
+        {:ws, _} ->
+          ConCache.dirty_delete(@messages, socket_id)
+
+        _ ->
+          {:error, :websocket_not_found}
+      end
+    end)
+  end
+
+  @spec close_connection(ws_socket_id | http_request_id) :: :ok
+  def close_connection(message_id) do
+    ConCache.delete(@messages, message_id)
+  end
```

Comment:
I see the same problem that there was in the previous PR: these have to be called explicitly so they don't get cleaned up if, e.g., the request/ws process crashes, leading to resource leaking

Reply:
*If* the process can crash without calling terminate, then yes. However I'm not yet convinced that is the case

Reply:
Implemented in #15